98-V100-Improvements-PaletteDesigner

tobitege opened the pull request:
Fixes #98 

- Rewrote Fast Filter functionality in MainForm, allowing different color value formats as filter values.
- Saving and restoring of the last used filter mode in user settings.
- Changing colors in either grid should now work more consistantly (2-way updates).
- Refactor PaletteMapper to enhance regex organization and add new methods for color path resolution. Introduced a dictionary for grouped regex patterns and improved grammar mapping logic for better maintainability and performance.
- Added progress bar after loading a theme for the color mapping process.
- Refactored existing methods for better readability and performance.
- Cleaned up code and improved overall structure for maintainability.

The biggest change is that after a theme is loaded (from file), the colors are mapped in a special process to the corresponding colorname in the rightmost grid. This is not via hardcoded mappings, but done via a mapping grammar, that should even be future proof for most cases, if/when a new enum would be introduced (taking, that the few naming rules are maintained).
This works for about 80% of all color enums. The process can take several seconds and displays a modal progress bar.

For testing: take any of the repo's community themes, convert it and load it. Then the result becomes more obvious in the grid.

tobitege committed:
(feat) Enhance Palette Designer with Fast Filter and Color Management Improvements

- Introduced Fast Filter functionality in MainForm, allowing users to filter colors by name or RGB values.
- Added support for saving and restoring the last used filter mode in user settings.
- Enhanced color management with new methods for setting and retrieving scheme colors.
- Refactored existing methods for better readability and performance.
- Updated UI elements to improve responsiveness during color table updates.
- Cleaned up code and improved overall structure for maintainability.

tobitege committed:
Enhanced PaletteMapper further

tobitege committed:
Added new regex patterns for AppButtonMenuDocs and FormHeader mappings in PaletteMapper, and updated related return statements for improved color handling.

tobitege committed:
Refactor PaletteMapper to enhance regex organization and add new methods for color path resolution. Introduced a dictionary for grouped regex patterns and improved grammar mapping logic for better maintainability and performance.
Added progress bar after loading a theme for the color mapping process.

tobitege committed:
Temp hide ColorFilterBtn, might get functionality later.

tobitege commented:
Working on one compile fix due to a local file revert....

tobitege committed:
MainForm: fixed calls to ApplyScheme

tobitege commented:
Err... need to fix some stuff again :(

tobitege committed:
MainForm: fix OnPaletteSchemeColorChanged for prop update

tobitege commented:
Fixed interim problem with color mapping a single value change.

tobitege commented:
Turns out, of course there are more issues to look into with color mapping.
I'm already working on further refinements and then have to refactor stuff down again.
Hoping to get this PR in reviewable state on the weekend. I was a bit too optimistic, apparently. 😬

tobitege committed:
Enhance MainForm and PaletteMapper with improved color management and performance optimizations

- Added a dictionary for mapping SchemeBaseColors to DataGridView row indices for faster updates.
- Implemented pre-caching of enum-to-path mappings to speed up color edits.
- Refactored UpdateGridRow method to utilize the new mapping for efficient row updates.
- Improved error handling in color property updates with exception logging.
- Updated regex patterns in PaletteMapper for better color path resolution and added new mappings for additional properties.

tobitege committed:
PaletteMapper/Grammar: more (final?) mapping fixes; fix line-endings in MainForm designer file

tobitege committed:
Mapping fixes; performance improvements.

- Added "Rebuild Palette Mapping" tool in MainForm for regenerating color mappings (cache file).
- Refactored palette initialization logic into a dedicated method for better code organization.
- Introduced fast compiled accessors in PaletteMapper for improved color retrieval performance.
- Code cleanup; better method summaries.

tobitege committed:
Refactored embedded controls page into its own Pages user control.

- Refactor win: main UI palette applies uniformly.
- Created a general instruction plan on how to refactor MainForm parts into standalone user control in Pages (for AI); six outstanding potential refactors.
- Integrated ControlsPage user control into the design for better organization.
- Updated palette application logic to include the new ControlsPage.

tobitege committed:
Refactor Grids into self-contained GridPage from MainForm.

- Removed KryptonNavigatorDesignGrids and associated logic from MainForm.
- Integrated gridPage1 for improved organization and encapsulation of grid functionality.
- Updated palette application logic to include new gridPage1.

tobitege committed:
Update PaletteMapper and project file for improved color mapping and output directory settings

- Added output directory setting for PaletteMapperCache.xml in the project file.
- Corrected color mapping paths in PaletteMapper for various UI elements to ensure accurate color retrieval.
- Enhanced color path resolution logic in PaletteMapperGrammar for better consistency and maintainability.

tobitege committed:
Refactor MainForm to enhance color filtering functionality and UI elements

- Removed obsolete ColorFilterBtn_SelectedColorChanged method to streamline color filtering logic.
- Introduced ColorFilterBtn_Click method for improved color selection via LiveColorPickerDialog.
- Updated MainForm.Designer.cs to replace ColorFilterBtn with KryptonButton and added ResetFilterBtn for better UI interaction.
- MainForm.resx: added new images for delete and color.
- Enhanced LiveColorPickerDialog to conditionally show extra controls based on the ShowExtraControls property.

tobitege commented:
This should do it for now for this PR from my side.

tobitege commented:
I'll wait for 2nd review before merging. 😁

giduac commented:
Hi @lesandrog,

Have you been testing this?
We would be interested in your findings...

lesandrog commented:
> Hi @lesandrog,
> 
> Have you been testing this? We would be interested in your findings...

I tried testing it but I'm getting some errors.
<img width="1096" height="527" alt="image" src="https://github.com/user-attachments/assets/d5fd4b81-80d2-4fba-9a08-3c5f5e97b8ec" />

tobitege commented:
@lesandrog - thanks for taking the time! Sorry to hear about the errors, but are you sure, that you have current alpha sources of the toolkit in access of the project?

lesandrog commented:
Now I've managed to compile.
@tobitege , when you change the theme, shouldn't the colors change (some of the buttons remain unchanged), and should the colors in your list be updated?
<img width="1902" height="831" alt="image" src="https://github.com/user-attachments/assets/3351655e-e9a8-44fd-b77e-b8089aaa8e74" />

Perhaps this option to change themes in the settings can be disabled!
And leave only the "New from Default Theme" option, which works correctly.

tobitege commented:
> Now I've managed to compile. @tobitege , when you change the theme, shouldn't the colors change (some of the buttons remain unchanged), and should the colors in your list be updated?
> 
> Perhaps this option to change themes in the settings can be disabled! And leave only the "New from Default Theme" option, which works correctly.

Yeh, I know what you mean, that was my thought when I started with it, too. 😃 
But the theme changer in the Settings is only meant for the UI itself, not the test/demo components. Ideally it should not affect those or the color grids at all. Otherwise you'd constantly have mixed appearances.

lesandrog commented:
@tobitege There's something strange with some themes.
MS365 Silver, for example: when you choose it through "New from Default Theme," the ribbon loads the correct colors, but during processing, it changes to darker colors.
![ezgif-8b1c8f7c12cae3](https://github.com/user-attachments/assets/24c80b51-c182-4c4c-9e61-4e31dadfb7b9)
<img width="938" height="664" alt="image" src="https://github.com/user-attachments/assets/84b1719e-1a74-4cde-9de9-7a6e53356dd0" />
<img width="908" height="638" alt="image" src="https://github.com/user-attachments/assets/afdc235d-b576-4481-a78a-4dd1a56ab2bb" />
<img width="900" height="645" alt="image" src="https://github.com/user-attachments/assets/2fdd805e-13d7-4925-be4d-66e25928de9a" />

tobitege commented:
Thanks @lesandrog , I'll investigate that.

tobitege committed:
Enhance MainForm and PaletteMapper with improved color handling and UI updates

- Updated MainForm to ensure palette colors are copied correctly from the base palette and removed redundant color copy calls.
- Enhanced color mapping logic in PaletteMapper to prefer file-provided values when fast path resolution fails.
- Improved grid updates in response to color changes to ensure UI reflects live updates accurately.

tobitege commented:
@lesandrog when you have time, please give it another try. I've disabled the processing for new themes for now.

lesandrog commented:
@tobitege, It's fine for me, I didn't find anything wrong.

giduac commented:
@tobitege & @lesandrog 

Are we getting to a stage where we can draw up some kind of guidelines / rules to aid in "repairing" the themes?

giduac commented:
Hi @lesandrog,

If this is OK with you, you can approve the PR...
